File: Objective-C/Reference App/Support/OutputViewController.m
//
//  OutputViewController.m
//  Reference App
//
//  Created by David Deller on 5/3/18.
//  Copyright © 2018 TallyGo. All rights reserved.
//

#import "OutputViewController.h"

@import TallyGoKit;
@import MapKit;

@interface OutputViewController ()

@property (nonatomic, weak) IBOutlet UITextView *outputView;

@end

@implementation OutputViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.outputView.layer.borderWidth = 0.5;
    self.outputView.layer.borderColor = [UIColor colorWithWhite:0.85 alpha:1].CGColor;
    self.outputView.layer.cornerRadius = 4;
}

#pragma mark -

- (void)handleOutput:(id)output {
    if ([output isKindOfClass:NSValue.class] && CLLocationCoordinate2DIsValid(((NSValue *)output).MKCoordinateValue)) {
        CLLocationCoordinate2D location = ((NSValue *)output).MKCoordinateValue;
        [self printOutput:[NSString stringWithFormat:@"%f,%f", location.latitude, location.longitude]];
        
    } else if ([output isKindOfClass:NSArray.class] && [((NSArray *)output).firstObject isKindOfClass:TGSearchResult.class]) {
        NSArray<TGSearchResult *> *searchResults = output;
        NSMutableString *resultsText = NSMutableString.new;
        for (TGSearchResult *searchResult in searchResults) {
            [resultsText appendString:searchResult.placeAddress];
            [resultsText appendString:@"\n"];
        }
        
        [self printOutput:[NSString stringWithString:resultsText]];
        
    } else if ([output isKindOfClass:TGSearchResult.class]) {
        TGSearchResult *searchResult = output;
        NSString *resultText = searchResult.placeAddress;
        
        [self printOutput:resultText];
        
    } else if ([output isKindOfClass:TGRoute.class]) {
        TGRoute *route = output;
        NSMutableString *routeText = NSMutableString.new;
        for (TGRouteSegment *segment in route.segments) {
            for (TGPoint *point in segment.points) {
                if (point.turn == nil) {
                    continue;
                }
                
                NSString *pointText = [NSString stringWithFormat:@"%@ on %@\n", point.turn.direction, point.turn.street.name];
                [routeText appendString:pointText];
            }
        }
        
        [self printOutput:[NSString stringWithString:routeText]];
        
    } else {
        [self handleErrorWithTitle:@"Unable to handle output format" message:nil viewController:nil];
    }
}

- (void)clearOutput {
    self.outputView.text = @"";
}

- (void)printOutput:(NSString *)output {
    if (output != nil) {
        self.outputView.text = output;
    } else {
        [self clearOutput];
    }
}

@end

File: Swift 4/Reference App/Support/OutputViewController.swift
//
//  ModalOutputViewController.swift
//  Reference App
//
//  Created by David Deller on 5/1/18.
//  Copyright © 2018 TallyGo. All rights reserved.
//

import UIKit
import TallyGoKit

class OutputViewController: ExampleViewController {
    
    @IBOutlet weak var outputView: UITextView!
    
    // MARK: -
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        outputView.layer.borderWidth = 0.5
        outputView.layer.borderColor = UIColor(white: 0.85, alpha: 1).cgColor
        outputView.layer.cornerRadius = 4
    }
    
    // MARK: -
    
    func handleOutput(_ output: Any) {
        if let location = output as? CLLocationCoordinate2D {
            printOutput("\(location.latitude),\(location.longitude)")
            
        } else if let searchResults = output as? [TGSearchResult] {
            let resultsText = (searchResults.map { (searchResult) -> String in
                return searchResult.placeAddress ?? ""
                
            } as NSArray).componentsJoined(by: "\n")
            
            printOutput(resultsText)
            
        } else if let searchResult = output as? TGSearchResult {
            let resultText = searchResult.placeAddress ?? ""
            
            printOutput(resultText)
            
        } else if let route = output as? TGRoute {
            let routeText: String = (route.segments.map({ (segment) -> String in
                return (segment.points.filter({ (point) -> Bool in
                    return (point.turn != nil)
                }).map({ (point) -> String in
                    return "\(point.turn!.direction.rawValue) on \(point.turn!.street?.name ?? "")"
                }) as NSArray).componentsJoined(by: "\n")
            }) as NSArray).componentsJoined(by: "\n")
            
            printOutput(routeText)
            
        } else {
            handleError(title: "Unable to handle output format")
        }
    }
    
    // MARK: -
    
    func clearOutput() {
        outputView.text = ""
    }
    
    func printOutput(_ output: String?) {
        if let output = output {
            outputView.text = output
        } else {
            clearOutput()
        }
    }

}
